From 0c4ffe2c48cbc962ebbdf83f342ff7eef521ee05 Mon Sep 17 00:00:00 2001
From: egg82 <phantom_zero@ymail.com>
Date: Tue, 25 Sep 2018 13:03:49 -0600
Subject: [PATCH] Add village API

This adds the Bukkit/API-facing code for a village API.
This API allows viewing and modifying of data specfically related to villages such
as number of golems or villagers, player reputation, village center and radius, etc.

All of this data is visible in NMS, but this API will expose some of the more useful methods for plugins.

diff --git a/src/main/java/org/bukkit/Village.java b/src/main/java/org/bukkit/Village.java
new file mode 100644
index 00000000..655dd7ba
--- /dev/null
+++ b/src/main/java/org/bukkit/Village.java
@@ -0,0 +1,115 @@
+package org.bukkit;
+
+// Paper - Add village API
+public interface Village {
+    /**
+     * Returns the location of the center of the village.
+     * 
+     * @return The location of the center of the village
+     */
+    Location getCenter();
+
+    /**
+     * Returns the radius of the village in blocks.
+     * 
+     * @return The radius of the village
+     */
+    int getRadius();
+
+    /**
+     * Returns the number of doors in the village.
+     * 
+     * @return The number of doors in the village
+     */
+    int getNumDoors();
+
+    /**
+     * Returns the numbers of villagers in the village.
+     * 
+     * @return The number of villagers in the village
+     */
+    int getNumVillagers();
+
+    /**
+     * Returns the number of villagers this village can hold.
+     * 
+     * @return The max villagers this village can hold
+     */
+    int getMaxVillagers();
+
+    /**
+     * Returns the number of golems in the village.
+     * 
+     * @return The number of golems in the village
+     */
+    int getNumGolems();
+
+    /**
+     * Returns the number of golems this village can hold.
+     * 
+     * @return The max golems this village can hold
+     */
+    int getMaxGolems();
+
+    /**
+     * Returns the amount of reputation this player has with the village.
+     * 
+     * @param player The player to check reputation for
+     * @return The amount of reputation this player has with the village
+     */
+    int getReputation(OfflinePlayer player);
+
+    /**
+     * Adds reputation to the player for this village. You may use negative values
+     * to remove reputation instead. The player's total reputation will then be
+     * clamped to a min and max, determined by the current server version's
+     * mechanics.
+     * 
+     * @param player The player to modify reputation for
+     * @param delta  The amount of reputation to modify
+     * @return The total reputation the player current has
+     */
+    int addReputation(OfflinePlayer player, int delta);
+
+    /**
+     * Resets reputation for every player that has visited the village to a new
+     * value.
+     * 
+     * @param newReputation The new reputation to reset to
+     */
+    void resetAllReputation(int newReputation);
+
+    /**
+     * Determines whether or not the specified player is considered hostile to the
+     * village. Iron golems from the village will attach the player.
+     * 
+     * @param player The player to check reputation for
+     * @return Whether or not the player is considered hostile to the village
+     */
+    boolean isHostile(OfflinePlayer player);
+
+    /**
+     * Returns true if there are no doors left in the village, rendering it
+     * annihilated.
+     * 
+     * @return True if there are no doors left
+     */
+    boolean isAnnihilated();
+
+    /**
+     * Returns true if the village is currently in a mating period, or false if not.
+     * 
+     * @return True if the village is in a mating period
+     */
+    boolean isMatingSeason();
+
+    /**
+     * Begins the village mating season.
+     */
+    void beginMatingSeason();
+
+    /**
+     * Ends the village mating season for a period of time.
+     */
+    void endMatingSeason();
+}
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 52b7bfec..b08d82b2 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1119,6 +1119,15 @@ public interface World extends PluginMessageRecipient, Metadatable {
     public Entity getEntity(UUID uuid);
     // Paper end
 
+    // Paper start - add village API
+    /**
+     * Returns a list of all villages in this world.
+     * 
+     * @return A list of all Villages in this world
+     */
+    public List<Village> getVillages();
+    // Paper end
+
     /**
      * Gets the unique name of this world
      *
-- 
2.19.0

